### PR TITLE
fix(rust, python): Fix is_in cannot cast list type for float

### LINKED ISSUE #10764: is_in broken with "cannot cast list type"
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://crates.io/crates/polars) of Polars.

### Reproducible example

issue affects both python and rust.

```python
import polars as pl

df = pl.DataFrame({
    "lst": [[1.0, 2.0, 3.0, 4.0, 5.0], [3.14, 5.28]],
    "n": [3.0, 2.0],
})

df.select(pl.col('n').is_in('lst'))
```

### Issue description

This issue affects both python and rust. I built latest `main` at c1487afa88a7d8c5930e8ec64a074b7d025107a9 — not yet released, would end up going into py-0.18.16 or rs-0.33.0 whever they go out. The currently released version seems to be working fine so it's a regression I stumbled into that has not yet been released. FYI @ritchie46 looks like you're actively developing around here.

### Expected behavior

expected behavior:
```
shape: (2, 1)
┌───────┐
│ n     │
│ ---   │
│ bool  │
╞═══════╡
│ true  │
│ false │
└───────┘
```

actual behavior:
```
File ~/scratch/polars/py-polars/polars/dataframe/frame.py:7663, in DataFrame.select(self, *exprs, **named_exprs)
   7561 def select(
   7562     self, *exprs: IntoExpr | Iterable[IntoExpr], **named_exprs: IntoExpr
   7563 ) -> DataFrame:
   7564     """
   7565     Select columns from this DataFrame.
   7566
   (...)
   7661
   7662     """
-> 7663     return self.lazy().select(*exprs, **named_exprs).collect(no_optimization=True)

File ~/scratch/polars/py-polars/polars/utils/deprecation.py:95, in deprecate_renamed_parameter.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
     90 @wraps(function)
     91 def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
     92     _rename_keyword_argument(
     93         old_name, new_name, kwargs, function.__name__, version
     94     )
---> 95     return function(*args, **kwargs)

File ~/scratch/polars/py-polars/polars/lazyframe/frame.py:1692, in LazyFrame.collect(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, no_optimization, slice_pushdown, comm_subplan_elim, comm_subexpr_elim, streaming)
   1680     comm_subplan_elim = False
   1682 ldf = self._ldf.optimization_toggle(
   1683     type_coercion,
   1684     predicate_pushdown,
   (...)
   1690     streaming,
   1691 )
-> 1692 return wrap_df(ldf.collect())

ComputeError: cannot cast List type (inner: 'Float64', to: 'Float64')
```

### Installed versions

on `main` branch at c1487afa88a7d8c5930e8ec64a074b7d025107a9, I used `make build` with python version 3.10.10 and latest stable rust (1.72.0).

### PR DESCRIPTION
**Pull request by reswqa**
This fixes #10764.

### CODE REVIEW

File: crates/polars-ops/src/series/ops/is_in.rs [hunk 1]
```diff
@@ -373,17 +373,35 @@ pub fn is_in(s: &Series, other: &Series) -> PolarsResult<BooleanChunked> {
...
         },
         DataType::Float32 => {
-            let other = other.cast(&DataType::Float32)?;
-            let other = other.f32().unwrap();
-            let other = other.reinterpret_unsigned();
+            let other = match other.dtype() {
+                DataType::List(_) => {
+                    let other = other.cast(&DataType::List(Box::new(DataType::Float32)))?;
+                    let other = other.list().unwrap();
+                    other.reinterpret_unsigned()
```
Comment (reswqa) [inline crates/polars-ops/src/series/ops/is_in.rs:380]: TBH, I'm not sure if this(`reinterpret list`) is on the right way. And I have a question: will it produce unexpected results if we directly cast to `List<UInt32>`?

Comment (ritchie46) [reply crates/polars-ops/src/series/ops/is_in.rs:380]: If we cast they will round up. Floating point values in a hash operations are iffy at best. I decided to only support them by interpreting their bytes as is. We also do this in group_by with floating values.

File: crates/polars-core/src/chunked_array/ops/bit_repr.rs [hunk 2]
```diff
@@ -165,6 +197,27 @@ impl Reinterpret for Float32Chunked {
     }
 }
 
+#[cfg(feature = "reinterpret")]
+impl Reinterpret for ListChunked {
+    fn reinterpret_signed(&self) -> Series {
+        match self.inner_dtype() {
+            DataType::Float32 => reinterpret_list_chunked::<Float32Type, Int32Type>(self),
+            DataType::Float64 => reinterpret_list_chunked::<Float64Type, Int64Type>(self),
+            _ => unimplemented!(),
```
[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-core/src/chunked_array/ops/bit_repr.rs:206]: I think `PolarsResult<ListChunked>`. But yeah, a result is preferred over a panic.

[the remark this answers is not in the document]
Comment (reswqa) [reply crates/polars-core/src/chunked_array/ops/bit_repr.rs:206]: Yes, `PolarsResult<ListChunked>` sounds more reasonable.

[the remark this answers is not in the document]
Comment (reswqa) [reply crates/polars-core/src/chunked_array/ops/bit_repr.rs:206]: `Reinterpret ` trait require a `Series` as return type, that's why I choose `PolarsResult<Series>`. Another approach is we introduce a new trait `ReinterpretList` for this case.

File: crates/polars-ops/src/series/ops/is_in.rs (hunk 1 above)
Comment (reswqa) [reply crates/polars-ops/src/series/ops/is_in.rs:380]: I see, make sense.

File: crates/polars-core/src/chunked_array/ops/bit_repr.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-core/src/chunked_array/ops/bit_repr.rs:206]: Ah, no. I think a `Series` is fine then. :+1:

**Comment by ritchie46** [comment]
Great fix! :)

**Pull request merged**